os/fs/smartfs/smartfs_utils.c

```diff
@@ -1015,232 +1076,259 @@ int smartfs_finddirentry(struct smartfs_mountpt_s *fs, struct smartfs_entry_s *d
 			entry = (struct smartfs_entry_header_s *)&fs->fs_rwbuffer[offset];
 		}
+		direntry->dsector = psector;
+		psector = nextsector;
+	}
 
-		/* If we found an entry, stop the search */
-
-		if (found) {
-			break;
-		}
-
-		/* If there are no more sectors, then we need to add one to make
-		 * room for the new entry.
-		 */
-
-		if (nextsector == SMARTFS_ERASEDSTATE_16BIT) {
-			/* Allocate a new sector and chain it to the last one */
-
-			ret = FS_IOCTL(fs, BIOC_ALLOCSECT, 0xFFFF);
-			if (ret < 0) {
-				goto errout;
-			}
-
-			nextsector = (uint16_t)ret;
-
-			/* Chain the next sector into this sector sector */
-
-			chainheader->nextsector[0] = (uint8_t)(nextsector & 0x00FF);
-			chainheader->nextsector[1] = (uint8_t)(nextsector >> 8);
-			//*((uint16_t *)chainheader->nextsector) = nextsector;
-			readwrite.offset = offsetof(struct smartfs_chain_header_s, nextsector);
-			readwrite.count = sizeof(uint16_t);
-			readwrite.buffer = chainheader->nextsector;
-			ret = FS_IOCTL(fs, BIOC_WRITESECT, (unsigned long)&readwrite);
-			if (ret < 0) {
-				fdbg("Error chaining sector %d\n", nextsector);
-				goto errout;
-			}
+errout:
+	if (ret != OK) {
+		return -ENOENT;
+	}
 
-#ifdef CONFIG_SMARTFS_USE_SECTOR_BUFFER
-			 /* TODO We will modify whole of create entry to integrity MTD operation soon.
-			  * We will always create sector first, and then update inside of sector. 
-			  * This will help to clear by fs_recover because all sector will become to isolated sector...
-			  * It is temp FIX, If we use sector buffer to support CRC, currently created new parent sector which is chained 
-			  * to previous parent sector is not committed yet. So Now we break loop to prevent read error 
-			  */
-			offset = sizeof(struct smartfs_chain_header_s);
-			entry = (struct smartfs_entry_header_s *)&fs->fs_rwbuffer[offset];
-			psector = nextsector;
-			break;
-#endif
-		}
+	return OK;
+}
 
-		/* Now update to the next sector */
+/****************************************************************************
+ * Name: smartfs_writeentry
+ *
+ * Description: Writes the new entry to the entry allocated by
+ *   find_availableentry or createentry.
+ *
+ * Input Parameters:
+ *   fs - pointer to smartfs mountpoint structure.
+ *   new_entry - the details of the entry in MTD where new entry is to be
+ *     written.
+ *   filename - name of new entry.
+ *   type - type of new entry (file/dir).
+ *   mode - mode of creation of new entry.
+ *   direntry - pointer to smartfs entry object to return details of newly
+ *     written entry.
+ *
+ * Returned Values:
+ *   OK - on successfully writing new entry to MTD.
+ *   ERROR - appropriate error code returned otherwise.
+ *
+ ****************************************************************************/
 
-		psector = nextsector;
-	}
+int smartfs_writeentry(struct smartfs_mountpt_s *fs, struct smartfs_entry_s new_entry, const char *filename, uint16_t type, mode_t mode, struct smartfs_entry_s *direntry, bool newchain)
```

Comment:
Your code flows below. (For example we create /mnt/a)
1. First it create firstsector by smartfs_alloc_firstsector. 
    It request allocsect to MTD layer, and keep returned target alloc sect.
    So now. newly written sector : 0 newly allocated sector : 1 (firstsector)

2. Now it try to find availableentry to get offset where target entry can be inserted.
    If there is space, fs->rw_buffer hold data of entire sector.
    So now. newly written sector : 0 newly allocated sector : 1 (firstsector)

3. If no space in parent sector, it alloc new parent sector and update previous sector's chain.
   Actually it should be done last time
    So now. newly written sector : 0 newly allocated sector : 2 (firstsector, 2nd parent sector - 1st sector has next sector value)

4. Finally it comes here.
    But if there was no space in 1st parent sector, newchain set as TRUE.
    In this case, this funtion never write first sector... so link error will be happened.
    so first you should write firstsector, and then if chained, then write 2nd parent sector.
    (and then write chain value of first sector is correct sequence)
   
 So this code can create only 12 files per directory.

Reply:
below is pseudo code.
```
		if (find_current_path_exist) {
			return -EEXIST;
		}
		
		smartfs_alloc_firstsector(); // Here we alloc first sector.
		
		Entry entry; //Entry which target to be inserted
		Entry parententry; //Information of parent entry;
		
		ret = find_availableentry(&entry, &parententry);
		if (ret == ENOENT) {
			/* Now we got new entry, also next parent sector's number if current parent is full */
			create_entry(&entry, &parententry); 
		} else {
			/* No next parent, now we found target entry & parententry's info */
			if (ret == OK) {
				fvdbg("found entry\n");
			} else {
				fdbg("error returned ret : %d errno : %d\n", ret, errno);
			}
		}

		/* Now we write first sector */
		if (sf) {
			sf->currsector = entry->firstsector;
			sf->parentsector = parentsector; //expand or reuse structure to represent it.
		} else {
			/* Below flow should be done by write_entry */
			write_sector(entry); // write first sector
		}
			
		/* Update chaining */
		if (chained) {
			if (sf) {
				sf->x = previous_sector_of_parent_sector;
			} else {
				write_next_parent(pareantentry);
				chain_parent(parententry);
			}
		}
```
We should expand sf for sync_internal...
```	
		sync_internal()
		
		write_sector(sf->currsector);
		write_sector(sf->parentsector);
		chain_parent(sf->parent, sf->x);
```

Reply:
**sf** buffer changes will be an improvement as discussed.
Will be done in a separate PR however.